### ISSUE TITLE
Difference between CharIn and Either

### ISSUE THREAD

**Issue by jtrakk** [gkappler/CombinedParsers.jl#22]
What is the difference between `CharIn` and `Either`? `CharIn` isn't restricted to `Char` types, so that might not be the best name, but maybe both `CharIn` and `Either` could be `OneOf`?

**Comment by gkappler** [comment]
I like your suggested simplification with `OneOf` but doubt whether eliminating one of the internal implementations is possible without cost -- even though there might be some code-cruft overlap in these types.

The reason is:
- `CharIn` is optimized for matching the element (Char) at current index in the sequence only (so requires match length 1 that might span multiple codeunits). A internal `ismatch` function (that can handle case e.g.) is used and state does not keep track of option #. 

- `Either` is very general: iterates through all matches of all options, can match longer subsequences, and retains the matching option # and state.


I agree that `CharIn` is not a good name when other sequences than `String`s are parsed. `OneOf` is catchy, but suggests a Set like representation, and I find it not matching in cases of a predicate function like `CharIn(islowercase)`. Would `ElementIn` clean up confusion for you?

**Comment by jtrakk** [comment]
> Would ElementIn clean up confusion for you?

Yeah, though it's a bit weird to say `ElementIn('x')` since `'x'` isn't a container. `Single('x')` is another option. Or maybe this is a case for `Just`.

**Comment by gkappler** [comment]
With `Just` I think of `Maybe` `Nothing` ;-).

You could say `parser('x')` in your example. 

`ElementIn('x', 'y')`, `ElementIn("xy")` make sense, yes?

Doe `ValueIn('x', 'y')`, `ValueIn("xy")` make more sense to you?

Another alternative would be to decide what implementation is used in the `Either` constructor?
